### ISSUE TITLE
0.8.0 - Too many open files

### ISSUE THREAD

**Issue by dsg-anthony** [promregator/promregator#188]
## Environment
Promregator 0.8.0
Cent OS 7
delivered jar 

## Summary / Problem Statement
File descriptors grow until promregator crashes.
java.io.IOException: Too many open files

## Observed Behavior
![image](https://user-images.githubusercontent.com/43544235/98894906-d8149a80-2473-11eb-94fc-271fd999bd5c.png)

```yaml
---
server:
  port: 50001
cf:
  api_host: ${promregator.url}
  username: ${promregator.username}
  password: ${promregator.password}
  cache.type: CAFFEINE
  connectionPool:
    size: 32
logging:
  level:
    org:
      cloudfoundry:
        promregator: ERROR
promregator:
  scraping:
    threads: 14
  authenticator:
    type: none
```

**Comment by checketts** [comment]
Thanks for opening this issue. Is this a new issue you are seeing with v0.8.0?

**Comment by dsg-anthony** [comment]
correct, 0.7.7 did not have this issue.

**Comment by eaglerainbow** [comment]
Hi @dsg-anthony,

Thanks for reporting on issue!
@tlkamp Do you see anything similar in your installation?

@dsg-anthony: It's strange that it "suddenly starts" to have the issue - and then almost scales linearly. It reminds me of our cf-java-client library resource leak, which @checketts had found out (and fixed). It could be super-helpful, if you could have a look at the open file descriptors (I assume you are running on a some sort of linux machine) and to what they refer (I have a certain suspicion there...).
You may do so by:
```bash
promregatorpid=`ps -ef | grep promregator.jar | awk '{print $2}'`
```

to find out on which processid (a.k.a. pid) promregator is running. Afterwards, you may list the open file descriptors with
```bash
ls -la /proc/$promregatorpid/fd
```
Under normal circumstances I would assume that you get there a list of ~50-80 files.  If you get more than 1000 lines, then it would be interesting to where these file descriptors all point to (right hand side of `->` printed there)...

Thanks for your analysis!

**Comment by eaglerainbow** [comment]
Note: might be related to https://github.com/cloudfoundry/cf-java-client/issues/1054#issuecomment-692598193

**Comment by tlkamp** [comment]
We do not have this issue. We are still running 0.8.0-RC-2 (have not had time to upgrade since the full release).

**Comment by eaglerainbow** [comment]
Note: upgrading cf-java-client to the 4.x.x series implies to upgrade to spring-boot 2.3.x. There is a cf-java-client 5.x.x series coming out soon, which will be based on spring-boot 2.4.x. 
On the 3.x.x series, there is a 3.29.x with some bugfixes included. Could be an option to try that first.

**Comment by dsg-anthony** [comment]
@eaglerainbow 
Here is a excerpt from `ls -la /proc/$promregatorpid/fd`
The full output of the command looks like this with few java lib files and /dev/random
![image](https://user-images.githubusercontent.com/43544235/99085083-c4f9eb80-2595-11eb-897d-e3f9f35376e5.png)

**Comment by dsg-anthony** [comment]
@eaglerainbow  I should also clarify in the graph I posted the increase is right after the .jar was updated to 0.8.8

**Comment by eaglerainbow** [comment]
Thanks, this fits to my suspicion. 
I am just trying to reproduce the file handle leak here in my local environment, but apparently I can't: I am rather steadily between 62 and 82 open files. I would say that this is very normal and does not cause me headache. 

On the other hand, [twoseat's comment](https://github.com/cloudfoundry/cf-java-client/issues/1054#issuecomment-726327258) causes me some headache: If even he is not able to determine the impact properly, then this must be something rather ugly (forget a second about this version track discussion that he brought up - that's yet another story).
In general I would sign the statement that we should be on the latest patch version of our consumable cf library track (changing the track to a newer track is another topic, though). That is why with 27419744768400d809c15a47cc9cbae11e4ac064 and 0e1810f29a8a12dfd7f8395f3d58f02f8d1a3050 I am now doing a little shot into the dark, hoping that this might improve your situation. 
The bad thing is that I am unable to properly test this here anywhere. If I created you a release candidate, could you test this in your environment?

**Comment by eaglerainbow** [comment]
Just for the record: 
* With 0.8.0-rc2, we were using reactor-netty 0.9.2
* With 0.8.0, we are running on reactor-netty 0.9.8
* With 0e1810f29a8a12dfd7f8395f3d58f02f8d1a3050, we would be running on reactor-netty 0.9.13

Note that the [patch notes of 0.9.9](https://github.com/reactor/reactor-netty/releases/tag/v0.9.9.RELEASE) states that there were file descriptor leaks fixed.
Note that 0.9.13 still has a regression issue which could netty-reactor make hang. That is only fixed with [0.9.14](https://github.com/reactor/reactor-netty/releases/tag/v0.9.14.RELEASE). However, cloudfoundry-client-reactor:3.29.0 only provides reactor-netty:0.9.13 :confounded:

**Comment by eaglerainbow** [comment]
I did some tests with 2b357bb63e5ab2891884ead37a5f089346d3ce18. Looking at the file descriptors, it appears that things are at least *better* - it took me less file descriptors to complete the entire test. However, the slope at the graph above and the slope I have observed locally are two different things. So, this all still is a little guess work...

@dsg-anthony: I'll create a release candidate for you soon, in the hope that you may find time and be able to check, if that solves your issue.

**Comment by eaglerainbow** [comment]
Here we go: [0.8.1-rc1](https://github.com/promregator/promregator/releases/tag/v0.8.1-rc1)

**Comment by dsg-anthony** [comment]
@eaglerainbow it looks like the issue is fixed with 0.8.1-rc1

![image](https://user-images.githubusercontent.com/43544235/99288976-5a051a80-280a-11eb-99f3-7a76471e2c2d.png)

**Comment by eaglerainbow** [comment]
>  it looks like the issue is fixed with 0.8.1-rc1 

Great to hear that! And the graph looks perfect to my mind...
I need to clarify, if the way of how I integrated reactor-netty:0.9.14 is sustainable. Either way, expect that you'll see a 0.8.1 soon...

**Comment by eaglerainbow** [comment]
I just released [0.8.2](https://github.com/promregator/promregator/releases/tag/v0.8.2). I had to skip 0.8.1 due to a technical problem in the release process (fixing that one would be a huge effort, compared to zero value-add).
Technically, this is the same thing as 0.8.1-rc1
